WordPress/gutenberg
# Migrate interface `enableItems` data to preferences package

Part of https://github.com/WordPress/gutenberg/issues/31965

## What?
Removes the final piece of `persistedState` from the interface package by migrating the `enableItems` data to the preferences store.

This feature was implemented in an interesting way, using some reducers called `singleEnableItems` (for `complementaryArea` state) and `multipleEnableItems` (for `pinnedItem` state) and it seems like the intention might have been for the `interface` package to be an abstract way of handling the state of different types of UI. In this PR I remove that abstraction and just store the data as `pinnedItems` and `complementaryArea` 😄 

## Why?
As outlined in #31965, the goal is to move persisted data out of our various stores to a centralized preferences store.

To explain a bit more about why I removed the `singleEnableItems` and `multipleEnableItems` concepts. I think `preferences` is already a generic data store, and so these abstractions no longer seem required. Other interfaces that need to persist state can do so by using the preferences store directly. It also simplifies how these features work.

## How?
Uses a migration function to migrate the persisted `localStorage` data. The selectors now select from and actions dispatch to the preferences store.

The `interface` package no longer has a reducer for its store, which is a bit weird, but the selectors and actions need to be kept around for backwards compatibility.

## Testing Instructions
I found the easiest way to test it to use my test environment and enable the 'Gutenberg Test Plugin, Plugins API' plugin. Then:
1. Checkout `trunk`
2. Visit the post editor and remove any WP local storage data
3. There should be two plugins available in the top right of the screen that have the same icon: 
![Screen Shot 2022-03-17 at 2 54 06 pm](https://user-images.githubusercontent.com/677833/158753435-eab7e0e4-0f08-4ddf-ba82-5c4edb047775.png)
4. Select one, and in the sidebar that opens, use the star icon to unpin it (the icon in the topbar should disappear)
5. Select the other plugin. Unpin it, but then pin it again. Make sure you leave this second plugin sidebar open and visible. 
6. Check out this branch, build it and reload the editor.
7. When the editor loads the first plugin (from step 4) shouldn't be pinned, the second plugin (from step 5) sidebar should still be open, visible and pinned (the localStorage state was migrated correctly)
8. Do some more pinning, unpinning and changing which sidebar is open on the right side of the screen
9. Reload and the state of those pinned items and visible sidebars should be persisted across reloads.
